jupyter_server/services/kernels/kernelmanager.py

```diff
@@ -320,37 +321,62 @@ async def restart_kernel(self, kernel_id):
         await ensure_async(self.pinned_superclass.restart_kernel(self, kernel_id))
         kernel = self.get_kernel(kernel_id)
         # return a Future that will resolve when the kernel has successfully restarted
-        channel = kernel.connect_shell()
+        shell_channel = kernel.connect_shell()
+        iopub_channel = kernel.connect_iopub()
         future = Future()
+        info_future = Future()
+        iopub_future = Future()
 
         def finish():
             """Common cleanup when restart finishes/fails for any reason."""
-            if not channel.closed():
-                channel.close()
             loop.remove_timeout(timeout)
             kernel.remove_restart_callback(on_restart_failed, 'dead')
 
         def on_reply(msg):
```

Comment:
This should be renamed to `on_shell_reply` as in https://github.com/jupyter/notebook/pull/5904.

Reply:
thx.